File: aspnet-core/modules/platform/LINGYUN.Platform.Domain/LINGYUN/Platform/Menus/Menu.cs
using JetBrains.Annotations;
using LINGYUN.Platform.Routes;
using System;
using Volo.Abp;

namespace LINGYUN.Platform.Menus;

/// <summary>
/// 菜单
/// </summary>
public class Menu : Route
{
    /// <summary>
    /// 框架
    /// </summary>
    public virtual string Framework { get; set; }
    /// <summary>
    /// 菜单编号
    /// </summary>
    public virtual string Code { get; set; }
    /// <summary>
    /// 菜单布局页,Layout的路径
    /// </summary>
    public virtual string Component { get; set; }
    /// <summary>
    /// 所属的父菜单
    /// </summary>
    public virtual Guid? ParentId { get; set; }
    /// <summary>
    /// 所属布局标识
    /// </summary>
    public virtual Guid LayoutId { get; set; }
    /// <summary>
    /// 公共菜单
    /// </summary>
    public virtual bool IsPublic { get; set; }
    protected Menu() 
    {
    }

    public Menu(
        [NotNull] Guid id,
        [NotNull] Guid layoutId,
        [NotNull] string path, 
        [NotNull] string name,
        [NotNull] string code,
        [NotNull] string component,
        [NotNull] string displayName,
        [NotNull] string framework,
        string redirect = "",
        string description = "",
        Guid? parentId = null,
        Guid? tenantId = null) 
        : base(id, path, name, displayName, redirect, description, tenantId)
    {
        Check.NotNullOrWhiteSpace(code, nameof(code));

        LayoutId = layoutId;
        Code = code;
        Component = component;// 下属的一级菜单的Component应该是布局页, 例如vue-admin中的 component: Layout, 其他前端框架雷同, 此处应传递布局页的路径让前端import
        Framework = framework;
        ParentId = parentId;

        IsPublic = false;
    }
}

File: vueJs/src/api/menu.ts
import ApiService from './serviceBase'
import { Route, PlatformType } from './layout'
import { urlStringify } from '@/utils/index'
import { ISortedResultRequest, PagedResultDto, PagedAndSortedResultRequestDto, ListResultDto } from './types'

const sourceUrl = '/api/platform/menus'
/** 远程服务地址 */
const serviceUrl = process.env.VUE_APP_BASE_API

/** 路由相关pi接口 */
export default class MenuService {
  public static get(id: string) {
    const _url = sourceUrl + '/' + id
    return ApiService.Get<Menu>(_url, serviceUrl)
  }

  public static getAll(payload: GetAllMenu) {
    const _url = sourceUrl + '/all?' + urlStringify(payload)
    return ApiService.Get<ListResultDto<Menu>>(_url, serviceUrl)
  }

  public static getList(payload: GetMenuByPaged) {
    const _url = sourceUrl + '?' + urlStringify(payload)
    return ApiService.Get<PagedResultDto<Menu>>(_url, serviceUrl)
  }

  public static getMyMenuList(platformType: PlatformType) {
    const _url = sourceUrl + '/by-current-user?platformType=' + platformType
    return ApiService.Get<ListResultDto<Menu>>(_url, serviceUrl)
  }

  public static getRoleMenuList(role: string, platformType: PlatformType) {
    const _url = sourceUrl + `/by-role/${role}/${platformType}`
    return ApiService.Get<ListResultDto<Menu>>(_url, serviceUrl)
  }

  public static getUserMenuList(userId: string, platformType: PlatformType) {
    const _url = sourceUrl + `/by-user/${userId}/${platformType}`
    return ApiService.Get<ListResultDto<Menu>>(_url, serviceUrl)
  }

  public static create(payload: MenuCreate) {
    return ApiService.Post<Menu>(sourceUrl, payload, serviceUrl)
  }

  public static update(id: string, payload: MenuUpdate) {
    const _url = sourceUrl + '/' + id
    return ApiService.Put<Menu>(_url, payload, serviceUrl)
  }

  public static delete(id: string) {
    const _url = sourceUrl + '/' + id
    return ApiService.Delete(_url, serviceUrl)
  }

  public static setRoleMenu(payload: RoleMenu) {
    const _url = sourceUrl + '/by-role'
    return ApiService.Put<void>(_url, payload, serviceUrl)
  }

  public static setUserMenu(payload: UserMenu) {
    const _url = sourceUrl + '/by-user'
    return ApiService.Put<void>(_url, payload, serviceUrl)
  }
}

export class MenuCreateOrUpdate {
  name!: string
  path!: string
  component!: string
  displayName!: string
  description?: string
  redirect?: string
  isPublic!: boolean
  platformType!: PlatformType
  meta: {[key: string]: any} = {}
}

export class MenuCreate extends MenuCreateOrUpdate {
  layoutId!: string
  parentId?: string
}

export class MenuUpdate extends MenuCreateOrUpdate {

}

export class GetAllMenu implements ISortedResultRequest {
  filter = ''
  sorting = ''
  reverse = false
  parentId?: string
  layoutId?: string
  platformType?: PlatformType
}

export class GetMenuByPaged extends PagedAndSortedResultRequestDto {
  filter = ''
  reverse = false
  layoutId?: string
  parentId?: string
  platformType?: PlatformType
}

export class Menu extends Route {
  code!: string
  layoutId!: string
  component!: string
  platformType!: PlatformType
  parentId?: string
  isPublic!: boolean
  children = new Array<Menu>()
}

export class RoleMenu {
  roleName!: string
  menuIds = new Array<string>()
}

export class UserMenu {
  userId!: string
  menuIds = new Array<string>()
}
